{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "76f717cb-6957-49eb-8cef-0ad53a5a0055",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter the fasta filename here: E_coli_genomic\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The entered file name does not exist. Please try again.\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Re-enter the fasta filename here: E_coli_genomic_DNA_seqs.fasta\n",
      "Set the minimal size of ORF (default: 50 bp):  150\n",
      "Enter the output filename for ORF analysis:  ORF_analysis_result.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The program can identify open reading frames from multiple DNA sequences\n",
      "--------------------------------------------------------------------------------\n",
      "The ORF analysis is done and the output result has been written into ORF_analysis_result.fasta.\n"
     ]
    }
   ],
   "source": [
    "# The OOP-based program for open reading frame analysis\n",
    "# Use E_coli_genomic_DNA_seqs.fasta as an input for program testing\n",
    "# ORF_analysis_result.fasta is the output from the program\n",
    "\n",
    "class ORF_analysis_program:\n",
    "    \n",
    "    # Initiate the class and define the class attributes for fasta, orf_size_threshold,\n",
    "    # and output filename\n",
    "    def __init__(self):\n",
    "        \n",
    "        import os\n",
    "        filename = input(\"Enter the fasta filename here:\").strip()\n",
    "        \n",
    "        while filename not in os.listdir():\n",
    "            print(\"The entered file name does not exist. Please try again.\")\n",
    "            filename = input(\"Re-enter the fasta filename here:\").strip()\n",
    "        \n",
    "        with open(filename) as file:\n",
    "            fasta = file.readlines()\n",
    "        \n",
    "        error = True\n",
    "        \n",
    "        while error:\n",
    "            try:\n",
    "                orf_size_threshold = int(input(\"Set the minimal size of ORF (default: 50 bp): \"))\n",
    "                if orf_size_threshold < 9:\n",
    "                    print(\"Your entered ORF size is less than 9 bp. Please try again.\")\n",
    "                    error = True\n",
    "                else:\n",
    "                    error = False\n",
    "            except:\n",
    "                print(\"No entry detected. Please re-enter the minimal size of ORF.\")\n",
    "        \n",
    "        output_filename = input(\"Enter the output filename for ORF analysis: \").strip()\n",
    "        \n",
    "        self.fasta = fasta\n",
    "        self.orf_size_threshold = orf_size_threshold\n",
    "        self.output_filename = output_filename\n",
    "\n",
    "    # Define a class method to generate a dictionary to store ID and sequence \n",
    "    # extracted from the fasta\n",
    "    def sequence_dictionary(self):\n",
    "        seq_dict = {}\n",
    "        for line in self.fasta:\n",
    "            if \">\" in line:\n",
    "                seq = \"\"\n",
    "                items = line.split()\n",
    "                seq_ID = items[0] + \"|\" + \" \".join([items[1], items[2], items[3]]) + \"|\"\n",
    "            elif line == \"\\n\":\n",
    "                seq_dict[seq_ID] = seq\n",
    "            else:\n",
    "                seq += line.strip().upper().replace(\" \", \"\")\n",
    "\n",
    "        self.seq_dict = seq_dict\n",
    "        return self.seq_dict\n",
    "    \n",
    "    # Define a class method to convert the DNA sequence into its reverse complement sequence\n",
    "    def reverse_complement(self, sequence):\n",
    "        nt_complement_dict = {'A':'T', 'C':'G', 'G':'C', 'T':'A'}\n",
    "        sequence_c = \"\"\n",
    "\n",
    "        for nt in sequence:\n",
    "            sequence_c += nt_complement_dict.get(nt)\n",
    "\n",
    "        sequence_rc = sequence_c[::-1]\n",
    "        return sequence_rc\n",
    "    \n",
    "    # Define a class method to split the DNA sequence into 3-nt codons from three different frames\n",
    "    def three_frames_codon_arrays(self, sequence):\n",
    "        frame_1 = []\n",
    "        frame_2 = []\n",
    "        frame_3 = []\n",
    "        three_frames = [frame_1, frame_2, frame_3]\n",
    "\n",
    "        for i in range(3):\n",
    "            for n in range(i, len(sequence)-2, 3):\n",
    "                codon = sequence[n:n+3]\n",
    "                three_frames[i].append(codon)\n",
    "            \n",
    "        return three_frames  \n",
    "    \n",
    "    \n",
    "    # Define a class method to generate codon arrays from 6 different reading frames \n",
    "    # (3 frames from each forward strand and reverse-complement strand sequences)\n",
    "    def six_frames_codon_arrays(self, sequence):\n",
    "        sequence_rc = self.reverse_complement(sequence)\n",
    "        frames_1 = self.three_frames_codon_arrays(sequence)\n",
    "        frames_2 = self.three_frames_codon_arrays(sequence_rc)\n",
    "        six_frames = frames_1 + frames_2\n",
    "        six_frames_dict = {}\n",
    "        for i in range(6):\n",
    "            six_frames_dict[i+1] = six_frames[i]\n",
    "        return six_frames_dict\n",
    "    \n",
    "    \n",
    "    # Define a class method to identify the open reading frames from a DNA sequence\n",
    "    def ORF_finder(self, frame_array, orf_size_threshold):\n",
    "        import re\n",
    "\n",
    "        start_codon = \"ATG\"\n",
    "        stop_codons = [\"TAA\", \"TAG\", \"TGA\"]\n",
    "        orf = \"\"\n",
    "        orf_ls = []\n",
    "\n",
    "        for codon in frame_array:\n",
    "            if codon == start_codon:\n",
    "                if re.search(r\"^ATG\", orf):\n",
    "                    orf += codon + \" \"\n",
    "                else:\n",
    "                    orf = \"\"\n",
    "                    orf += codon + \" \"\n",
    "            elif codon not in stop_codons:\n",
    "                orf += codon + \" \"\n",
    "            else:\n",
    "                orf += codon\n",
    "                orf_ls.append(orf)\n",
    "                orf = \"\"\n",
    "\n",
    "        orf_list = []\n",
    "        for orf in orf_ls:\n",
    "            if re.search(r\"^ATG\", orf):\n",
    "                orf_list.append(orf)\n",
    "\n",
    "        orf_list_select = []\n",
    "        for orf in orf_list:\n",
    "            if len(orf.replace(\" \", \"\")) >= orf_size_threshold:\n",
    "                orf_list_select.append(orf)\n",
    "\n",
    "        return orf_list_select   \n",
    "    \n",
    "    \n",
    "    # Define a class method to generate a dictionary to store all ORFs found from a sequence\n",
    "    def ORF_dictionary(self, sequence, orf_size_threshold):\n",
    "        orf_dict = {}\n",
    "        six_frames_dict = self.six_frames_codon_arrays(sequence)\n",
    "        for frame_num, frame_array in six_frames_dict.items():\n",
    "            orf_list_select = self.ORF_finder(frame_array, orf_size_threshold)\n",
    "            orf_dict[frame_num] = orf_list_select\n",
    "        return orf_dict\n",
    "    \n",
    "    \n",
    "    # Define a class method to generate a dictionary to store all ORFs found from multiple sequences\n",
    "    def ORF_dictionary_multi_seqs(self, orf_size_threshold=50):\n",
    "        orf_dict_multi_seq = {}\n",
    "        for ID, sequence in self.seq_dict.items():\n",
    "            orf_dict = self.ORF_dictionary(sequence, orf_size_threshold)\n",
    "            orf_dict_multi_seq[ID] = [sequence, orf_dict]\n",
    "        return orf_dict_multi_seq\n",
    "    \n",
    "    \n",
    "    # Define a function to return gencode dictionary\n",
    "    def gencode(self):\n",
    "        codons_dict = {\n",
    "                       'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "                       'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "                       'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "                       'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "                       'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "                       'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "                       'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "                       'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "                       'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "                       'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "                       'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "                       'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "                       'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "                       'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "                       'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "                       'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W'}\n",
    "        \n",
    "        self.codons_dict = codons_dict        \n",
    "        return self.codons_dict\n",
    "    \n",
    "    \n",
    "    # Define a function to translate ORF into a protein sequence\n",
    "    def protein_translation(self, orf):\n",
    "        codons_dict = self.gencode()\n",
    "        codons = orf.split()\n",
    "        protein = \"\"\n",
    "        for codon in codons:\n",
    "            if codons_dict.get(codon) != \"_\":\n",
    "                protein += codons_dict.get(codon)\n",
    "            else:\n",
    "                break\n",
    "        return protein\n",
    "    \n",
    "    \n",
    "    # Define a class method to write identified ORFs from all six reading frames into a file\n",
    "    def ORF_analysis_output(self, orf_dictionary):\n",
    "        import re\n",
    "        output = open(self.output_filename, \"w\")\n",
    "        for ID, seq_orf_dict in orf_dictionary.items():\n",
    "            sequence = seq_orf_dict[0]\n",
    "            orf_dict_ = seq_orf_dict[1]\n",
    "            for num, orfs in orf_dict_.items():\n",
    "                if num < 4:\n",
    "                    for orf in orfs:\n",
    "                        orf_ns = orf.replace(\" \", \"\")\n",
    "                        length = len(orf_ns)\n",
    "                        start = re.search(orf_ns, sequence).start() + 1\n",
    "                        end = re.search(orf_ns, sequence).end()\n",
    "                        title = f\"{ID}FRAME={num} POS={start} LEN={length}\"\n",
    "                        output.write(title + \"\\n\")\n",
    "                        for n in range(0, len(orf), 60):\n",
    "                            output.write(orf[n:n+60] + \"\\n\")\n",
    "                        output.write(\"\\n\")\n",
    "                        \n",
    "                        protein = self.protein_translation(orf)\n",
    "                        protein_id = f\">The translated protein sequence, {len(protein)} amino acids.\"\n",
    "                        output.write(f\"{protein_id}\\n\")\n",
    "                        for a in range(0, len(protein), 60):\n",
    "                            output.write(protein[a:a+60] + \"\\n\")\n",
    "                        output.write(\"\\n\")\n",
    "                        \n",
    "                else:\n",
    "                    for orf in orfs:\n",
    "                        orf_ns = orf.replace(\" \", \"\")\n",
    "                        length = len(orf_ns)\n",
    "                        sequence_rc = self.reverse_complement(sequence)\n",
    "                        start = re.search(orf_ns, sequence_rc).start() + 1\n",
    "                        end = re.search(orf_ns, sequence_rc).end()\n",
    "                        title = f\"{ID}FRAME={num} POS={-(start)} LEN={length}\"\n",
    "                        output.write(title + \"\\n\")\n",
    "                        for n in range(0, len(orf), 60):\n",
    "                            output.write(orf[n:n+60] + \"\\n\")\n",
    "                        output.write(\"\\n\")\n",
    "                        \n",
    "                        protein = self.protein_translation(orf)\n",
    "                        protein_id = f\">The translated protein sequence, {len(protein)} amino acids.\"\n",
    "                        output.write(f\"{protein_id}\\n\")\n",
    "                        for a in range(0, len(protein), 60):\n",
    "                            output.write(protein[a:a+60] + \"\\n\")\n",
    "                        output.write(\"\\n\")\n",
    "                        \n",
    "        output.close()\n",
    "        \n",
    "    \n",
    "    # Define a main() method to execute the entire ORF analysis program\n",
    "    def ORF_analysis_main(self):\n",
    "        \n",
    "        print(\"The program can identify open reading frames from multiple DNA sequences\")\n",
    "        print(\"-\" * 80)\n",
    "        \n",
    "        self.sequence_dictionary()\n",
    "        orf_dict_multi_seq = self.ORF_dictionary_multi_seqs(self.orf_size_threshold) \n",
    "        self.ORF_analysis_output(orf_dict_multi_seq)\n",
    "        \n",
    "        print(f\"The ORF analysis is done and the output result has been written into {self.output_filename}.\")\n",
    "        \n",
    "\n",
    "# Assign the program class and call the main method for ORF \n",
    "ORF_analysis = ORF_analysis_program()\n",
    "ORF_analysis.ORF_analysis_main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb7c3a2f-6e33-4b5e-8be2-3f245b8287b6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
